server/serverAPI/controller/ShoppingCartController.ts

```diff
@@ -0,0 +1,248 @@
+                await this.verifySchema(quantitySchema, res);
+                ingredientSchema.quantity = quantitySchema;
+            } catch (e) {
+                return Promise.reject(e);
+            }
+        }
+
+        return ingredientSchema;
+    }
+
+    /**
+     * Returns all food in user shopping list.
+     * 
+     * @param req Request parameter that holds information about request.
+     * @param res Response parameter that holds information about response.
+     */
+    getAll = async (req: Request, res: Response) => {
+        let parameters = new Map<string, any>([["username", req.serverUser.username]]);
+
+        try {
+            let user = await this.requestGet(parameters, res)
+            return this.sendSuccess(200, res, user.shoppingCart);
+        } catch (e) {
+            return e;
+        }
+    }
+
+    /**
+     * Adds food to user's shopping list.
+     * 
+     * @param req Request parameter that holds information about request.
+     * @param res Response parameter that holds information about response.
+     */
+    add = async (req: Request, res: Response) => {
+        let parameters = new Map<string, any>([["username", req.serverUser.username]]);
+
+        try {
+            let user = await this.requestGet(parameters, res);
+
+            let ingredientSchema = await this.parseAddRequest(req, res);
+
+            let exists: boolean = false;
+
+            for (let i = 0; i < user.shoppingCart.length; i++) {
+                let existingItem = user.shoppingCart[i];
+
+                if (existingItem.id === ingredientSchema.id &&
+                    existingItem.recipeID === ingredientSchema.recipeID
+                ) {
+                    exists = true;
+
+                    if (existingItem.recipeID !== null) {
+                        return this.sendError(400, res, "Ingredient for this recipe already exists in shopping cart.");
+                    }
+
+                    let amount = ingredientSchema.quantity;
+
+                    if (existingItem.quantity.unit !== ingredientSchema.quantity.unit) {
+                        let convertedUnit = await this.foodAPI.ConvertUnits(ingredientSchema.quantity, existingItem.quantity.unit, existingItem.name);
+                        
+                        if (convertedUnit === null) {
+                            return this.sendError(400, res, "Amount units cannot be converted.");
+                        }
+
+                        amount = convertedUnit;
+                    }
+
+                    user.shoppingCart[i].quantity.value += amount.value;
+
+                    break;
+                }
+            }
+
+            if (!exists) {
+                user.shoppingCart.push(ingredientSchema);
+            }
+
+            let updatedUser = await this.requestUpdate(req.serverUser.username, user, res);
+            return this.sendSuccess(200, res, updatedUser.shoppingCart);
+        } catch (e) {
+            return e;
+        }
```

Comment:
What's the difference between returning vs. throwing an error?

Reply:
returning works... throw doesn't :rofl: